Fix the index creation error on table created through SELECT INTO

pranavJ23 opened the pull request:
### Description
Issue - 
If a table is formed via SELECT INTO command then the column name does not get converted to lowercase.
Due to this, indexation over that column fails which is the issue we want to resolve.
This happens as the SELECT INTO command does not convert the column name in lowercase(save it as it is) whereas the index creation command search for the column name(over which indexation is done) in lower case.
### Issues Resolved
Conversion of column name to lower case(if column name contains uppercase character/s) in Select Into query.

Note: if the table is pre-existing then user needs to drop the table and create it again to enable this feature.
### Approach
1. Down-case the column name using downcase_identifier.
2. check via filter that resjunk should be false and the column name should not be NULL.

### Task
[BABEL-4926]
Signed-off-by: Pranav Jain pranavjc@amazon.com

### Test Scenarios Covered ###
1. Null case i.e if the column name is null.
3. Case when multiple column names are given in SELECT INTO query
4. Case when column name is already in lowercase 
5. Case when column name is passed with a different name in SELECT INTO statement(using as).
6. Case when column is passed via some function.
7. Case when column/s is/are passed via *(select all) instead of specifying names.
8. Case when column with special letters like $,& is passed into square brackets.
9. Case when identity function is used with select into statement.
10. Added mvu test when the select into is executed in prepare step and the create index is executed in verify step
11. Case when column is modified via ALTER TABLE or ALTER COLUMN 
12. Creating UNIQUE INDEX on the column after SELECT INTO
13. Dropping column via ALTER TABLE...DROP COLUMN after SELECT INTO
14. Case when table is passed to SELECT INTO via Table-Returning Function 
15. Multibyte Column Name (Chinese)
16. Multibyte Column Name (Chinese) more then 64 characters
17. Table with multi-part names
18. Table with multi-part names and column name length > 64

### Check List
- [Pranav Jain] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Index creation on table getting passed if table is created via SELECT INTO command, converted the col_name to lowercase and storing the original name as well

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Index creation on table passed if table is created via SELECT INTO command with capital column name

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Index creation on table if table is created via SELECT INTO command with
column name contains uppercase characters

Commit:
Index creation on table if table is created via SELECT INTO command with
column name contains uppercase characters
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Index creation on table if table is created via SELECT INTO command with column name contains uppercase characters
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6090,11 +6096,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 				tle->resno = current_resno;
 				modifiedTargetList = lappend(modifiedTargetList, tle);
+
+				if (original_name!=NULL&&strcmp(original_name,tle->resname))
+				{
```
nit: indentation, spaces between operator, variables

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6090,11 +6096,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+				}
 			}
 		}
+		
```
lets remove not required changes

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6090,11 +6096,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 			}
 			else
 			{
+				char* original_name = tle->resname;
```
only set original_name when `tle->resname!=NULL`

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6122,9 +6153,13 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 	if (altstmt)
 		result = lappend(result, altstmt);
 
+	if (tempStore!=NIL)
```
lets use a more descriptive variable name

tanscorpio7 commented on test/JDBC/expected/BABEL-4926.out:
```diff
@@ -0,0 +1,187 @@
+create index IDX_REPRODUCTION on BABEL_4926_REPO(numb,Numb1,ColuM);
+go
+-- output is the lowercase name,original_name
+select attname,attoptions from pg_attribute where attrelid = (select oid from pg_class where relname = 'babel_4926_repo');
```
also prefix column name with jira name. So that we can only select only those column here.

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4926

Commit:
Curretly Babelfish fails with index creation on table fails if table is created via SELECT INTO command but after this change if our column name contains the uppercase character then too the SELECT INTO command does not fails

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Curretly Babelfish fails with index creation on table fails if table is created via SELECT INTO command but after this change if our column name contains the uppercase character then too the SELECT INTO command does not fails

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

roshan0708 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6057,6 +6058,7 @@ static List *
 transformSelectIntoStmt(CreateTableAsStmt *stmt)
 {
 	List *result;
+	List *babel_4926_original_name_list;
```
Let's not use jira numbers in the actual code, use some other generic name for the variables.

forestkeeper replied:
This naming looks confusing

roshan0708 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6066,6 +6068,7 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 	into = stmt->into;
 	result = NIL;
 	altstmt = NULL;
+	babel_4926_original_name_list = NIL;
```
+1

roshan0708 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6159,11 +6162,40 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 			}
 			else
 			{
+				AlterTableStmt *newstmt;
+				AlterTableCmd *cmd;
+				char* original_name = NULL;
+				newstmt = NULL;
```
Try to stick to either CamelCase or SnakeCase when naming variables or functions. And try to make variable names meaningful in the current context.

roshan0708 commented on test/JDBC/jdbc_schedule:
```diff
@@ -7,9 +7,7 @@
 # 5. To add a test, add test name (without extension Eg. BABEL-TEST) on a
```
Fix: remove this file change since you're not adding anything new in the schedule file.

roshan0708 commented on test/JDBC/input/BABEL-4926.sql:
```diff
@@ -0,0 +1,158 @@
+-- create a new table BABEL_4926 1
```
Let's change the test file name. Use a name that is more suited to the fix.

roshan0708 commented on test/JDBC/input/BABEL-4926.sql:
```diff
@@ -0,0 +1,158 @@
+go
+
+
+
```
Refactor: Try to group tests as per the case, this makes it easier to review and check the edge cases.

rishabhtanwar29 commented on test/JDBC/jdbc_schedule:
```diff
@@ -7,9 +7,7 @@
 # 5. To add a test, add test name (without extension Eg. BABEL-TEST) on a
 #    new line
 # 6. If you want the framework to not run certain files, use: ignore#!#<test name>
-
 all
-
```
Unnecessary change.

roshan0708 commented:
Fix: add test file name in `upgrade/latest/schedule` file.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -98,6 +98,7 @@ extern bool pltsql_recursive_triggers;
 extern bool restore_tsql_tabletype;
```
[General] Please follow commit description guidelines, we are missing `Task` and `signoff` fields. Use description of https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2748 as an example.

roshan0708 reviewed (changes requested):
Fix: add test file name in `upgrade/latest/schedule` file.

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6191,9 +6223,13 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 	if (altstmt)
 		result = lappend(result, altstmt);
 
+	if (babel_4926_original_name_list != NIL)
```
I think this additional variable is not needed, we can directly append into `result` if needed at line number `6195`.
[Bonus] In fact we can take this opportunity to even remove the redundant `altstmt` variable as well and directly append into `result` whenever needed.

pranavJ23 replied:
I tried this but It was throwing an error, we are first appending the SelectInto statement commands and after that appending the altstmt or newstmt. if we directly append it to result then the newstmt will be in queue before selectInto commands and it will throw an error while executing(it won't find the new table that we have created using the select into statement)

rishabhtanwar29 commented on test/JDBC/expected/BABEL-4926.out:
```diff
@@ -0,0 +1,225 @@
+-- create a new table BABEL_4926 1
```
1. Let's have the meaningful test filenames which reflects the purpose instead of plain Task/jira numbers, something like `TEST-SELECT_INTO`, although feel free to have even a better name.
2. Let's have test cases in upgrade format: <testname>-vu-prepare/verify/cleanup.
3. In fact, we can use the existing test files for select-into (BABEL_539-vu-*) to add new test scenarios instead of creating separate test files. See https://github.com/babelfish-for-postgresql/babelfish_extensions/commit/1ece7f885dba6379f775ee15213a86f570d88b72

pranavJ23 replied:
done

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6159,11 +6162,40 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+				if (tle->resname != NULL)
+				{
+					original_name = tle->resname;
+					tle->resname = downcase_identifier(tle->resname, strlen(tle->resname), false, false);
```
Can this `tle-resname` be truncated if the name is larger than 64 chars? If yes then we are going to store truncated original name which is not desirable. Let's check it once, otherwise we will need to fetch the original_name from query string same as how we generally do it.

rishabhtanwar29 commented on test/JDBC/expected/BABEL-4926.out:
```diff
@@ -0,0 +1,225 @@
+create index IDX_REPRODUCTION on BABEL_4926_REPO(babel_4926_COL);
+go
+-- output is the lowercase name,original_name
+select attname,attoptions from pg_attribute where attname like '%babel_4926%' and attrelid = (select oid from pg_class where relname = 'babel_4926_repo') order by attname asc;
```
Not fancy enough ;-), let's come up with a query having join between pg_attribute and pg_class. Moreover, we don't want full `attoptions` list here, instead the query can be modified to just print the `bbf_original_name` element of that list.

[BONUS] Let's see if we can use `information_schema.columns` view instead for this purpose in which case you can ignore the suggestion above.

rishabhtanwar29 commented on test/JDBC/expected/BABEL-4926.out:
```diff
@@ -0,0 +1,225 @@
+~~START~~
+int
+~~END~~
+
```
Let's also add a tests with column names having length larger than 64 chars, special characters, quoted with `[]` and `""` etc.

rishabhtanwar29 commented on test/JDBC/input/BABEL-4926.sql:
```diff
@@ -0,0 +1,158 @@
+go
+-- dropping the table BABEL_4926_REPO
+drop table BABEL_4926_REPO;
+go
```
Nit: please add a newline at end of each new file.

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6159,11 +6162,40 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 			}
 			else
 			{
+				AlterTableStmt *newstmt;
```
We should add else if () ?? some condition to enter this block here, we can't make all the else conditions into this block actually.

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -98,6 +98,7 @@ extern bool pltsql_recursive_triggers;
 extern bool restore_tsql_tabletype;
 extern bool babelfish_dump_restore;
 extern bool pltsql_nocount;
+extern const char *ATTOPTION_BBF_ORIGINAL_NAME;
```
Can you show me a a const define case using Upper naming only ?

pranavJ23 replied:
This is an existing variable I am using

forestkeeper commented:
Please polish the description and make those points clear : 
1. what's the root cause ( Upper case/Lower case issue ? But more details pls )
2. How do we fix this ( for example : change the behavior of select into in certain step)
3. Test cases design ( we already have that, that's good )
4. What if the table is already created by previous wrong behavior, what will happen when create index on those tables ? ( Or we already covered it ? )

forestkeeper reviewed (changes requested):
Pls add mvu test when the select into is executed in prepare step and the create index is executed in verify step

Commit:
changed the tesst files and added column length constraint

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
changed the test file types and added column length constraint

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Index creation on table if table is created via SELECT INTO command with
column name contains uppercase characters

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

Commit:
Adding the test files in latest schedule
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
adding the SELECT_INTO_TEST to upgrade versions
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
change in the selecct_into output for mvu test
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
removing modifications from jdbc_schedule and adding test case in
upgrade latest schedule
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
fixing the issue when column_name_length>=64 original name will not be
stored
Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
empty

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4926

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6181,11 +6184,40 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 				tle->resno = current_resno;
 				modifiedTargetList = lappend(modifiedTargetList, tle);
+
+				if (original_name != NULL && strcmp(original_name, tle->resname) && strlen(original_name) == NAMEDATALEN)
+				{
```
`strlen(original_name) < NAMEDATALEN`
only then we save the original name.

tanscorpio7 replied:
actually truncation makes the name < NAMEDATALEN. so we may not be able to tell if a name is truncated or not just by looking at the length.

Commit:
removing the 64 length check

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
changing the original name storing condition (when column name >=64)

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6181,11 +6184,40 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 				tle->resno = current_resno;
 				modifiedTargetList = lappend(modifiedTargetList, tle);
+
+				if (original_name != NULL && strcmp(original_name, tle->resname))
```
Pls add comments about what this if block trying to do?

Commit:
fixing the case when column_name>64

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4926

Commit:
to handle the column_name_length>63, modifying the hooks.c and storing
the original name there only if select into is called and truncating it
later.

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
continuing changes via get_tle_by_resno

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
handling for identity function, column_name length <63

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
minor changes

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
Merge branch 'BABEL_4_X_DEV' into BABEL-4926

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6180,6 +6185,38 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+					AlterTableStmt *newstmt;
+					AlterTableCmd *cmd;
+
+					newstmt = NULL;
```
We can remove this line.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -98,6 +98,8 @@ extern bool pltsql_recursive_triggers;
 extern bool babelfish_dump_restore;
 extern bool pltsql_nocount;
+extern const char *ATTOPTION_BBF_ORIGINAL_NAME;
+extern TargetEntry *get_tle_by_resno(List *tlist,AttrNumber resno);
```
we can remove this line

Commit:
addressing comments

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6261,11 +6294,14 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 			}
 			else
 			{
+				
 				current_resno += 1;
 				tle->resno = current_resno;
 				modifiedTargetList = lappend(modifiedTargetList, tle);
+				
 			}
 		}
+		
```
nit: not needed change.

tanscorpio7 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6180,6 +6183,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 		foreach (elements, q->targetList)
 		{
 			TargetEntry *tle = (TargetEntry *)lfirst(elements);
+			char* original_name = NULL;
+
+			if (tle->resname != NULL && !tle->resjunk && strlen(tle->resname) < NAMEDATALEN -1)
```
Nit: spacing `char *original_name` &  `NAMEDATALEN - 1`

Commit:
addressing the comments

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6180,6 +6183,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+			{	
+
+				original_name = tle->resname;
+				tle->resname = downcase_identifier(tle->resname, strlen(tle->resname), false, false);
```
resname should be downcased irrespective of whether we can find original_name or not or in other words, down-casing should not get affect by whether or not `strlen(tle->resname) < NAMEDATALEN - 1`

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6180,6 +6183,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+				original_name = tle->resname;
+				tle->resname = downcase_identifier(tle->resname, strlen(tle->resname), false, false);
+
+				// check if the original_name is already in lowercase
```
Use PG-style comments: /* XXX */

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6293,9 +6328,13 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 	}
 
 	result = lappend(result, stmt);
```
My initial comment was to move this line up in the top and directly append everything else into `result` instead of creating new lists.

pranavJ23 replied:
it was not work, as altstmt should be appended before our column list. Altstmt create the identity column and our alterTablestmt(newstmt, which stores the original name) should be appended after Altstmt (the alterTablestmt which creates the identity column).

rishabhtanwar29 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,304 @@
+create index IDX_REPRODUCTION on select_into_REPO(select_into_COL);
+go
+-- output is the lowercase name,original_name
+select attname,attoptions from pg_attribute join pg_class on pg_attribute.attrelid=pg_class.oid where pg_class.relname='select_into_repo' and attname like '%select_into%'order by attname asc;
```
Have you considered using `information_schema.columns` instead?

pranavJ23 replied:
We want to test both downcased name and original name for the column whereas we will only get the original name from information_schema.columns

rishabhtanwar29 replied:
information_schema.columns returns either original_name or stored name depending upon whether or not the original_name exists.

rishabhtanwar29 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,304 @@
+select select_into_COL_select_into_COL_select_into_COL_select_into_COL_ into select_into_REPO from select_into;
+go
+-- Indexing over column select_into_COL
+create index IDX_REPRODUCTION on select_into_REPO(select_into_COL_select_into_COL_select_into_COL_select_into_COL_);
+go
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: column "select_into_col_select_into_colb68a72f6d624cf699ece5bf8f6446ce4" does not exist)~~
```
This shouldn't be failing.

pranavJ23 replied:
here the column name length > 63, so we are not lowercasing the column name for this case and not storing the original name

tanscorpio7 replied:
We want to lowercase the name for greater than 64 as well 
but not store the original name.

pranavJ23 replied:
Done

Commit:
addressing comment to lowercase teh column name even if column name length > 63

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
addressing comments

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6180,6 +6183,36 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
+				tle->resname = downcase_identifier(tle->resname, strlen(tle->resname), false, false);
+
+				/* check if the original_name is already in lowercase */
+				if (original_name != NULL && strcmp(original_name, tle->resname) && strlen(tle->resname) < NAMEDATALEN - 1)
```
I suspect `strlen(tle->resname) < NAMEDATALEN - 1` might differ for multi-bytes characters, let's check with @Deepesh125 regarding it.

Commit:
not storing original name

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
indentation changes

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
NULL handling

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

rishabhtanwar29 commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -6294,7 +6295,7 @@ transformSelectIntoStmt(CreateTableAsStmt *stmt)
 	result = lappend(result, stmt);
 	if (altstmt)
-		result = lappend(result, altstmt);
+		result = lappend(result, altstmt);	
```
Nit: unnecessary change.

Commit:
address comment

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

Commit:
Trigger Build

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

Deepesh125 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,300 @@
+go
+drop table if exists select_into_pre_exist_repro;
+go
+-- column length >=64 todo
```
with multibyte characters also, from Japanese, Chinese, etc

Deepesh125 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,300 @@
+select select_into_COL into select_into_REPO from select_into;
+go
+-- Indexing over column select_into_COL
+create index IDX_REPRODUCTION on select_into_REPO(select_into_COL);
```
Let's also add test cases to test 
1. ALTER TABLE...ALTER COLUMN works fine on this column
2. Add new constraint through ALTER TABLE ... ADD CONSTRAINT
3. CREATE UNIQUE INDEX
4. ALTER TABLE ... DROP COLUMN

Deepesh125 commented on test/JDBC/input/SELECT_INTO_TEST-vu-prepare.sql:
```diff
@@ -0,0 +1,12 @@
+create table select_into_pre_exist(select_into_pre_exist_COL int);
+go
+-- create a new function to double the value
+CREATE FUNCTION dbo.select_into_double_function (@in INT) RETURNS INT
```
Test with table returning function also

Deepesh125 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,300 @@
+
```
why aren't we testing any sys or information_schema_tsql views anywhere? How are we making sure that sys.column (for example) is showing correct details

Deepesh125 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -0,0 +1,300 @@
+go
+~~START~~
+varchar
+select_into_col_select_into_colb68a72f6d624cf699ece5bf8f6446ce4
```
just trying to understand, where is this truncation (ident + MD5) is happening when we are passing `truncate = false` to `downcase_identifier`?

pranavJ23 replied:
tle->resname is already truncated and MD5 file is already attached to it

Commit:
Adding test cases

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Deepesh125 commented on test/JDBC/expected/SELECT_INTO_TEST-vu-verify.out:
```diff
@@ -298,3 +298,242 @@ go
+
+
+-- Indexing over column 中文_COL
+create index IDX_REPRODUCTION on select_into_REPO(中文_COL);
```
with length > 64 also

Commit:
multibyte character with column name length >64

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Deepesh125 commented on test/JDBC/input/SELECT_INTO_TEST-vu-verify.sql:
```diff
@@ -0,0 +1,399 @@
+
```
Just realised that we missed adding tests cases for following scenarios - 
1. When source table's column length is >=64 and/or contains multi-byte characters
2. When source table's column is specified as multi-part, say, `select master.dbo.table.xyz into ...` , `select dbo.table.xyz into ...` , etc

Commit:
addressing comments to add testcases

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

Commit:
addressing test cases comments

Signed-off-by: Pranav Jain <pranavjc@amazon.com>

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

forestkeeper commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -561,5 +561,6 @@ reverse
 sys_sequences
 replicate
 space
+SELECT_INTO_TEST
```
why we only add this into latest ?

pranavJ23 committed:
Merge branch 'babelfish-for-postgresql:BABEL_4_X_DEV' into BABEL-4926

Commit:
adding SELECT_INTO_TEST to 15_9 schedule file

Signed-off-by: Pranav Jain <pranavjc@amazon.com>
